jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/CHManyToManyShortestPaths.java

```diff
@@ -0,0 +1,496 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.jgrapht.alg.shortestpath.ContractionHierarchy.ContractionEdge;
+import static org.jgrapht.alg.shortestpath.ContractionHierarchy.ContractionVertex;
+
+/**
+ * Efficient algorithm for the many-to-many shortest paths problem based on contraction hierarchy.
+ *
+ * <p>
+ * The algorithm is originally described in the article: Sebastian Knopp, Peter Sanders,
+ * Dominik Schultes, Frank Schulz, and Dorothea Wagner. 2007. Computing many-to-many shortest
+ * paths using highway hierarchies. In Proceedings of the Meeting on Algorithm Engineering &amp; Expermiments.
+ * Society for Industrial and Applied Mathematics, Philadelphia, PA, USA, 36-45.
+ *
+ * <p>
+ * First a backward single source shortest paths search is performed for each target vertex. During
+ * the searches a bucket $b(v)$ is associated with each vertex $v$ in the graph. A bucket stores
+ * a set of pairs $(t,d)$, where $t$ is a target vertex current search is performed from and $d$ is the computed
+ * distance from $v$ to this target. Then a forward single source shortest paths search is performed
+ * from every source vertex. When a search settles a vertex $v$ with distance $d(s,v)$, where $s$ is current
+ * source vertex, its bucket is scanned. For each entry $(t,d)$ if $d(s,t) &gt; d(s,v) + d$ values of paths weight
+ * between $s$ and $t$ and its middle vertex is updated. The middle vertices are then used to restored
+ * actual path from the information in the shortest paths trees.
+ *
+ * <p>
+ * The efficiency of this algorithm is derived from the fact that contraction hierarchy produces fairly small
+ * shortest paths trees. This allows to both speedup the computations and decrease memory usage to store the
+ * paths.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Semen Chudakov
+ */
+public class CHManyToManyShortestPaths<V, E> extends BaseManyToManyShortestPaths<V, E> {
+    /**
+     * The underlying graph.
+     */
+    private final Graph<V, E> graph;
+
+    /**
+     * Contraction hierarchy for {@code graph}.
+     */
+    private Graph<ContractionVertex<V>, ContractionEdge<E>> contractionGraph;
+    /**
+     * Mapping from vertices in the original {@code graph} to vertices
+     * in the {@code contractionGraph}.
+     */
+    private Map<V, ContractionVertex<V>> contractionMapping;
+
+
+    /**
+     * Constructs an instance of the algorithm for a given {@code graph}.
+     *
+     * @param graph a graph
+     */
+    public CHManyToManyShortestPaths(Graph<V, E> graph) {
+        this.graph = graph;
+        Pair<Graph<ContractionVertex<V>, ContractionEdge<E>>, Map<V, ContractionVertex<V>>> p
+                = new ContractionHierarchy<>(graph).computeContractionHierarchy();
+        this.contractionGraph = p.getFirst();
+        this.contractionMapping = p.getSecond();
```

Comment:
```java
Pair<Graph<ContractionVertex<V>, ContractionEdge<E>>, Map<V, ContractionVertex<V>>> p
                  = new ContractionHierarchy<>(graph).computeContractionHierarchy();
this(graph, p.getFirst(), p.getSecond())
```

Reply:
It is not possible to make this change, because `this` statement should be at the beginning of a constructor.